topy.py

```diff
@@ -140,41 +143,101 @@ def server(command: str):
+# so that disabled server won't receive any meesage and update its list after the time.
+# our state.neighbors list looks like this
+#  [[thisserverID, neighborserverID, cost], [thisserverID, neighborserverID, cost], [thisserverID, neighborserverID, cost], .....]
+
+def disable(state: Server_State, command:str):
+    # we are on server1, disable <serverid>
+    # ex) disable 3
+    #check if that id is really our neighbor
+    # if so
+    dstServerID = command[1]
+    state.routing_table[dstServerID]['cost'] = 'inf'
+    state.updatedIDs.append(dstServerID)
+    # remove serverid(ex: 3) from state.neighbors list
+    for item in state.neighbors:
+        if item[1] == dstServerID:
+            state.neighbors.remove(item)
 
-def disable():
-    pass
 
 def crash():
     pass
 
 def exit_func():
     pass
 
-def send_message(state: Server_State, ip: str, port: int):
+# Create message
+def formMessage(state: Server_State):
+    updatedIDs = (set(state.updatedIDs))
+    nFields = len(updatedIDs)
+    payload = {}
+    
+    #create header message
+    header = { "header": { "n_update_fields": nFields, "server_ip": state.ip, "server_port": state.port } }
+
+    #create payload message
+    server_response_temp = []
+    for item in updatedIDs:
+        if item in state.routing_table.keys():
+            cost = state.routing_table[item]['cost']
+            for server in state.servers:
+                if item == server[0]:
+                    server_response_temp.append({'ip': server[1], 'port': server[2], 'id': server[0], 'cost': cost})
+
+    payload["payload"] = server_response_temp
+    header.update(payload)
+    message = header
+    return(message)
+
+# Send routing table update to neighbors right away. Then reset updated list and timer.
+def step(state: Server_State):
+    if len(state.updatedIDs) != 0:
+        message = formMessage(state)
+        for neighbor in state.neighbors:
+            for server in state.servers:
+                if server[0] == neighbor[1]:
+                    neighborIP = server[1]
+                    neighborPort = server[2]
+                    send_message(message, neighborIP, neighborPort)
+        #reset updateIDs list after step
+        state.updatedIDs.clear()
+    else:
+        print('routing table has not been updated. we are not sending routing table.')
 
-    payload = json.dump(state.routing_table).encode('utf-8')
+def send_message(message, ip: str, port: int):
 
+    # used json.dumps instead of json.dump. It was throwing error with 'fp'
+    # payload = json.dumps(state.routing_table).encode('utf-8')
+    payload = json.dumps(message).encode('utf-8')
+    
     with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
-        s.connect((ip, port))
+        s.connect((ip, int(port)))
         #TODO need to know what to send
         s.sendall(payload)
 
+# update routing table with new cost and the list of updated servers
```

Comment:
the update function needs to be able to change any 2 given serverID's cost between the two

`server-ID1, server-ID2: The link for which the cost is being updated. ` from the assignment

Reply:
`this  command  will  be  issued  to  both  server-ID1  and  server-ID2  and  involve  them  to 
update the cost and no other server`